src/clusterfuzz/_internal/bot/tasks/setup.py

```diff
@@ -508,63 +521,126 @@ def update_fuzzer_and_data_bundles(fuzzer_name):
   if fuzzer.max_testcases and fuzzer.max_testcases < max_testcases:
     environment.set_value('MAX_TESTCASES', fuzzer.max_testcases)
 
-  # Check for updates to this fuzzer.
-  version_file = os.path.join(fuzzer_directory, '.%s_version' % fuzzer_name)
-  if (not fuzzer.builtin and
-      revisions.needs_update(version_file, fuzzer.revision)):
-    logs.log('Fuzzer update was found, updating.')
-
-    # Clear the old fuzzer directory if it exists.
-    if not shell.remove_directory(fuzzer_directory, recreate=True):
-      logs.log_error('Failed to clear fuzzer directory.')
-      return None
-
-    # Copy the archive to local disk and unpack it.
-    archive_path = os.path.join(fuzzer_directory, fuzzer.filename)
-    if not blobs.read_blob_to_disk(fuzzer.blobstore_key, archive_path):
-      logs.log_error('Failed to copy fuzzer archive.')
-      return None
+  # If the fuzzer generates large testcases or a large number of small ones
+  # that don't fit on tmpfs, then use the larger disk directory.
+  if fuzzer.has_large_testcases:
+    testcase_disk_directory = environment.get_value('FUZZ_INPUTS_DISK')
+    environment.set_value('FUZZ_INPUTS', testcase_disk_directory)
 
-    try:
-      archive.unpack(archive_path, fuzzer_directory)
-    except Exception:
-      error_message = ('Failed to unpack fuzzer archive %s '
-                       '(bad archive or unsupported format).') % fuzzer.filename
-      logs.log_error(error_message)
-      fuzzer_logs.upload_script_log(
-          'Fatal error: ' + error_message, fuzzer_name=fuzzer_name)
-      return None
-
-    fuzzer_path = os.path.join(fuzzer_directory, fuzzer.executable_path)
-    if not os.path.exists(fuzzer_path):
-      error_message = ('Fuzzer executable %s not found. '
-                       'Check fuzzer configuration.') % fuzzer.executable_path
-      logs.log_error(error_message)
-      fuzzer_logs.upload_script_log(
-          'Fatal error: ' + error_message, fuzzer_name=fuzzer_name)
-      return None
-
-    # Make fuzzer executable.
-    os.chmod(fuzzer_path, 0o750)
-
-    # Cleanup unneeded archive.
-    shell.remove_file(archive_path)
-
-    # Save the current revision of this fuzzer in a file for later checks.
-    revisions.write_revision_to_revision_file(version_file, fuzzer.revision)
-    logs.log('Updated fuzzer to revision %d.' % fuzzer.revision)
 
-  _clear_old_data_bundles_if_needed()
+def preprocess_update_fuzzer_and_data_bundles(fuzzer_name):
```

Comment:
It's getting hard to tell what runs where. Can we perhaps introduce a decorator "@trusted" or "@uworker" (or both?) that indicates this (and can do some form of runtime verification as well)?

Reply:
I'll start doing this. But I think we need to hold off on runtime verification because until every task is migrated functions like these will be run trusted and untrusted workers.